### PR TITLE
171 code refactoring predict

### LINKED ISSUE #171: Code Refactoring: Predict
# Issue description

## Type of Issue

Refactor the predict.py to make it easier to understand and add more docs

- [ ] Documentation update
- [ ] Test request
- [ ] Linting request

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

Fixes #171 

## Type of change

- [x] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by mathpluscode** [comment]
@YipengHu @NMontanaBrown this PR is almost ready, i.e. it runs now, i just need to add some tests to cover the funcs ;)

So please have a quick look first to see if there's anything obviously wrong

**Comment by YipengHu** [comment]
> @YipengHu @NMontanaBrown this PR is almost ready, i.e. it runs now, i just need to add some tests to cover the funcs ;)
> 
> So please have a quick look first to see if there's anything obviously wrong
> 
> I will update a resume on what the predict is doing tmr.
> 
> Otherwise, run the output test, and check the saved files under `logs`

looks good to me so far. if you add a simple/brief description of what the predict does now, i can add to docs.

**Comment by mathpluscode** [comment]
So, now predict saves the outputs for each pair of data,
- if the data is labeled, the folder will be named as `pair_xxx_label_yyy`
- if the data is unlabeled, the folder will be named as `pair_xxx`

Here we split the different labels for the same pair of images because:
depending on the algorithm, the labels can be used as inputs and therefore the outputs might depend on the labels.

Then inside the directory, we will save (if available)
- moving_image
- moving_label
- fixed_image
- fixed_label
- ddf
- dvf
- pred_moving_image (warped moving image)
- pred_fixed_label

in two forms
- nifti, `moving_image.nii.gz` inside the directory
- png, by depth under a sub-directory using the tensor's name, like `moving_image/depth0_moving_image.png`

except ddf and dvf, all other images are saved using gray colormap.

You can also check the outputs of the tests to see more details.

@YipengHu

**Comment by YipengHu** [comment]
@s-sd and/or @zacbaum i'm running out of time for this one, but do not want to delay things. I had a look before the tests were in. So if either of you is happy with it, please approve. Thank you!

**Comment by YipengHu** [comment]
> So, now predict saves the outputs for each pair of data,
> 
> * if the data is labeled, the folder will be named as `pair_xxx_label_yyy`
> * if the data is unlabeled, the folder will be named as `pair_xxx`
> 
> Here we split the different labels for the same pair of images because:
> depending on the algorithm, the labels can be used as inputs and therefore the outputs might depend on the labels.
> 
> Then inside the directory, we will save (if available)
> 
> * moving_image
> * moving_label
> * fixed_image
> * fixed_label
> * ddf
> * dvf
> * pred_moving_image (warped moving image)
> * pred_fixed_label
> 
> in two forms
> 
> * nifti, `moving_image.nii.gz` inside the directory
> * png, by depth under a sub-directory using the tensor's name, like `moving_image/depth0_moving_image.png`
> 
> except ddf and dvf, all other images are saved using gray colormap.
> 
> You can also check the outputs of the tests to see more details.
> 
> @YipengHu
> 
> I will add another issue, once this is merged to request for documentation page update.

What is pred_fixed_label? is it warped moving labels? should it be consistently named with pred_moving_image?

File: test/unit/test_train.py [hunk 1]
```diff
@@ -87,19 +54,28 @@ def test_build_callbacks():
         assert isinstance(callback, tf.keras.callbacks.Callback)
 
 
-def test_train():
+def test_train_and_predict():
```
Comment (s-sd) [inline test/unit/test_train.py:57]: should this maybe test if it produces an output in the logs instead of just if it is able to run?

File: test/unit/test_util.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    # labeled and have pred_moving_image
+    got = calculate_metrics(
+        fixed_image=fixed_image,
+        fixed_label=fixed_label,
+        pred_moving_image=pred_moving_image,
+        pred_fixed_label=pred_fixed_label,
+        fixed_grid_ref=fixed_grid_ref,
+        sample_index=sample_index,
+    )
+    assert got["image_ssd"] is not None
```
Comment (s-sd) [inline test/unit/test_util.py:169]: for the `is not None` statements is there a way to check the output instead?

**Comment by s-sd** [comment]
> @s-sd and/or @zacbaum i'm running out of time for this one, but do not want to delay things. I had a look before the tests were in. So if either of you is happy with it, please approve. Thank you!

@YipengHu I've had a look at the added tests and have left a few comments. 

@mathpluscode I think some additions might benefit the tests if they're possible (left as comments)

File: test/unit/test_util.py (hunk 2 above)
Comment (mathpluscode) [reply test/unit/test_util.py:169]: In line 147, I assumed the metric functions are correct, so checking the output numerically is not my objective here. what do you think, is it ok/reasonable?

File: test/unit/test_train.py (hunk 1 above)
Comment (mathpluscode) [reply test/unit/test_train.py:57]: You mean, test if there's an output folder?
For `train`, no need as the `predict` will have to use the checkpoint, for `predict` ok, i can check this ;)

**Comment by mathpluscode** [comment]
ooops @s-sd i forgot to rename the `pred_moving_image` to `pred_fixed_image` will do so and request approval again, or i will ask @YipengHu directly

**Comment by s-sd** [comment]
@mathpluscode, so the warped fixed image is saved alongside its label?

**Comment by mathpluscode** [comment]
@s-sd yes!
